proposed/0006-unified-structured-data-imporvement.md

```diff
@@ -0,0 +1,177 @@
+}
+```
+ 
+**step 4:**
+Client-A wants to transfer ownership of Structured Data to Client-B:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 5:**
+Logic: Since `previous_owner_keys` in incoming data contains something, use `key` = `(sha512(previous_owner_keys), Name)` for look-up. Calculate a new `key` = `(sha512(owner_keys), Name)` if the update is legal as checked below (as usual).
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 3 to verify signature of data in step 4. If valid replace previous data and store new one:
+```
+Key = (sha512(B-PublicKey), Name)
+Value = StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 6:**
+Client-A tries to update illegally:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: Anything,
+    previous_owner_keys: Anything/NONE,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 7:**
+If look-up succeeds, Vault will use owner field from **step 5** == `B-PublicKey` and signature verification fails. So it will not update anything.
+ 
+**step 8:**
+Client-B tries to update:
+```
+StructuredData {
+    other_fields: Value-33,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: NONE,
+    signatures: Signed by B-PrivateKey,
+}
+```
+ 
+**step 9:**
+Logic: Since `previous_owner_keys` in incoming data contains nothing, use `key` = `(sha512(owner_keys), Name)` for look-up.
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 5 to verify signature of data in step 8. If valid replace previous data and store new one:
+```
+Key = (sha512(B-PublicKey), Name)
+Value = StructuredData {
+    other_fields: Value-33,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: NONE,
+    signatures: Signed by B-PrivateKey,
+}
+```
+## Change in the Definition of GETs
+The above shows the cases for PUTs, POSTs and DELETEs. There will be a change in the signature of GETs to achieve the above no-collision scheme. GETs for StructuredData need to be changed from:
+```
+GET(Name, DataRequest::StructuredData(u64))
+```
+to
+```
+GET(Name, DataRequest::StructuredData(SHA512(PublicKeys)))
+```
+For ordinary cases where `data-PUTers` themselves are fetching thier data (majority of clients) this does not mean much as the client engine will inject this silently as it already has the knowledge of user public-keys. The only time it does not have the knowledge is during the fetching of the session packet. The solution for this is easy as login packet has a dedicated maidasafe-type-tag and vaults can recognise it and make special allowance for it to be stored with `Key = Name` instead of `Key = (sha512(PublicKeys), Name)` as in above.
+
```

Comment:
Please define such maidsafe-type-tag as a specific value. and is such maidsafe-type-tag is the type_tag field in structured_data?